In [7]:
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Social_Network_Ads.csv')
dataset.head(20)

User ID  Gender  Age  EstimatedSalary  Purchased
0   15624510    Male   19            19000          0
1   15810944    Male   35            20000          0
2   15668575  Female   26            43000          0
3   15603246  Female   27            57000          0
4   15804002    Male   19            76000          0
5   15728773    Male   27            58000          0
6   15598044  Female   27            84000          0
7   15694829  Female   32           150000          1
8   15600575    Male   25            33000          0
9   15727311  Female   35            65000          0
10  15570769  Female   26            80000          0
11  15606274  Female   26            52000          0
12  15746139    Male   20            86000          0
13  15704987    Male   32            18000          0
14  15628972    Male   18            82000          0
15  15697686    Male   29            80000          0
16  15733883    Male   47            25000          1
17  15617482    Male   45            26000          1
18  15704583    Male   46            28000          1
19  15621083  Female   48            29000          1

In [2]:
X = dataset.iloc[:, 2:4].values.astype(np.float64)
y = dataset.iloc[:, 4].values.astype(np.float64)

# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test =  train_test_split(X,y,test_size = 0.25, random_state = 54)

# Feature Scaling

from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

The Data is now pre-processed

Buiding KNN Model

In [3]:
def Dist(A,B):
    Sum=0
    for i in range(len(A)):
        Sum+=(A[i] - B[i])**2
    return Sum**(1/2)

def KNN(X_train , Y_train , point , k):
    import operator
    distn = []
    
    for i in range(len(X_train[:,0])):
        distn.append((Dist(X_train[i,:],point) , i))
        distn.sort(key = operator.itemgetter(0))
    
    Sum=0
    for i in range(k):
        Sum+=Y_train[distn[i][1]]
    if Sum/k > 0.5:
        return 1
    else:
        return 0
        

    
def predict(X_train , y_train , X_test,k):
    y_predict = np.zeros((len(X_test[:,0]),1))
    for i in range(len(X_test[:,0])):
        
        y_predict[i,0] = KNN(X_train , y_train , X_test[i,:],k)
    return y_predict

In [4]:
y_pred = predict(X_train , y_train , X_test,5)
pd.DataFrame(y_pred)

0
0   1.0
1   0.0
2   0.0
3   0.0
4   0.0
5   0.0
6   0.0
7   0.0
8   0.0
9   1.0
10  0.0
11  0.0
12  0.0
13  0.0
14  0.0
15  1.0
16  1.0
17  1.0
18  1.0
19  1.0
20  1.0
21  0.0
22  0.0
23  0.0
24  0.0
25  0.0
26  0.0
27  1.0
28  0.0
29  0.0
..  ...
70  1.0
71  0.0
72  0.0
73  1.0
74  0.0
75  0.0
76  0.0
77  0.0
78  0.0
79  0.0
80  0.0
81  1.0
82  1.0
83  0.0
84  1.0
85  1.0
86  1.0
87  0.0
88  0.0
89  0.0
90  1.0
91  1.0
92  0.0
93  0.0
94  0.0
95  0.0
96  0.0
97  0.0
98  1.0
99  0.0

[100 rows x 1 columns]

In [5]:
pd.DataFrame(y_test)

0
0   1.0
1   0.0
2   0.0
3   0.0
4   0.0
5   0.0
6   0.0
7   0.0
8   0.0
9   1.0
10  0.0
11  0.0
12  0.0
13  0.0
14  0.0
15  1.0
16  1.0
17  0.0
18  1.0
19  1.0
20  1.0
21  0.0
22  0.0
23  0.0
24  1.0
25  0.0
26  0.0
27  1.0
28  0.0
29  0.0
..  ...
70  1.0
71  0.0
72  0.0
73  1.0
74  0.0
75  0.0
76  0.0
77  0.0
78  0.0
79  0.0
80  0.0
81  1.0
82  1.0
83  0.0
84  1.0
85  1.0
86  1.0
87  0.0
88  0.0
89  0.0
90  1.0
91  1.0
92  0.0
93  0.0
94  0.0
95  0.0
96  0.0
97  0.0
98  1.0
99  0.0

[100 rows x 1 columns]

Assesing Model

In [6]:
# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
cm

array([[61,  3],
       [ 4, 32]])

Number of correct predictions = sum of left diagonal = 93

Number of incorrect predictions = sum of right diagonal = 7

Accuracy = 93%